### PR TITLE
fix: various fixes for the Run > Files view

### LINKED ISSUE #28: [View > Files > External Data] nothing shows up
**Steps to Reproduce**
1. Deploy and register LIGO Tale
1. *Tale Dashboard* > *Public Tales*: click "View" on "LIGO Tutorial"
1. Navigate to *Files* > *External Data*

**Expected Results**
See the content of the Tale's `dataSet`.

**Actual Results**
Panel is empty.

### LINKED ISSUE #30: [View > Files > External Data] (+) button is not disabled for readonly Tale
**Steps to Reproduce**
1. Deploy and register LIGO Tale
1. *Tale Dashboard* > *Public Tales*: click "View" on "LIGO Tutorial"
1. Navigate to *Files* > *External Data*

**Expected Results**
`(+)` button is grayed out.

**Actual Results**
`(+)` button is working and opens up modals

**NOTES**
It also applies to *Files* > *Workspace*. I'm not sure if I should file it as a separate bug, there's a weird behavior, when you try to upload a file into read-only workspace. It initially appears as successful, goes away after reload (which is expected), but most importantly it shows up in users Home directory. That suggest that something is really wrong with "parentId" and "parentType" that used in the request.

### LINKED ISSUE #31: [View > Files > External Data] Add from Catalog modal doesn't display currrent dataSet
**Steps to Reproduce**
1. Deploy and register LIGO Tale
1. *Tale Dashboard* > *Public Tales*: click "Run Tale" on "LIGO Tutorial" and click "Confirm"
1. *Tale Dashboard* > *My Tales*: click on "View" on "LIGO Tutorial" (your copy)
1. *View* > *Files* > *External Data*: click on `(+)`
1. Select *WT Data Catalog* from the dropdown

**Expected Results**
See current dataset in the right subpanel:
panel_good

**Actual Results**
Right subpanel is completely missing:
panel_bad

### LINKED ISSUE #33: [registration modal] Dropdown should auto select first result
**Steps to Reproduce**
1. Create a new Tale
2. Navigate to the new Tale's *View* > *Files* > *External Data*
3. Hit '(+)' and select *Web (DOI or URL)*
4. Enter "doi:10.5065/D6862DM8" in the field and click "Search"

**Expected Results**
One result is found, it's autoselected in the dropdown and auxiliary fields (repository, size etc) are populated

**Actual Results**
You need to click on the dropdown to select the only result.

**NOTES**
In a very unlikely, if not impossible, scenario that more than one item is found, we can grab first from the list.

### LINKED ISSUE #37: Downloading files that were uploaded to Home/Workspace works only after reload
**Steps to Reproduce**
1. Create a Tale
1. Navigate to *View* > *Files* > *Home*.
1. Using `(+)` upload a file.
1. Using action menu on the file, select and click "Download"

**Expected Results**
File downloads

**Actual Results**
File doesn't download. (Not traceback in the console)

**NOTES**
Download works after I refresh the page.

### LINKED ISSUE #41: Creating a folder in Home results with a new folder in workspace
**Steps to Reproduce**
1. Create a Tale
1. Navigate to *View* > *Files* > *Home*.
1. Using `(+)` create a folder "test".
1. Refresh the page

**Expected Results**
Folder `test` is still there.

**Actual Results**
Folder `test` is gone. Navigate to *View* > *Files* > *Workspace* and find a folder name `test`.

**NOTES**
I think that's the same thing I mentioned in #30 `parentId` and `parentType` are clearly wrong here.

### LINKED ISSUE #48: Upload file results in 400 bad request
**Steps to Reproduce**
1. Select Home or Workspace
2. Select "+" > Upload file
3. Upload a file

**Expected Results**
File is uploaded without error

**Actual Results**
File appears to be uploaded, but with 400 errors in alerts and console
```
Http failure response for https://girder.local.wholetale.org/api/v1/file/chunk?uploadId=5e95fd17f91daec313abd10e&offset=0: 400 Bad Reques
```

and

```
{"message": "Invalid upload id (5e95fd67f91daec313abd11a).", "type": "rest"}
```

### LINKED ISSUE #50: Adding external data results in 400
**Steps to Reproduce**
1. Create new Tale
2. Select Files> External Data > + > Web (DOI or URL)
3. Register doi:10.5065/D6862DM8
4. Select External Data > + > WT Data Catalog
5. Try to check  "Humans and Hydrology..." > usco2005.xls
 
**Expected Results**
File usco2005.xls is selected and added to external data

**Actual Results**
PUT 400 ``{"message": "Invalid JSON object for parameter tale: '_modelType' is a required property", "type": "rest"}``

### LINKED ISSUE #89: Files upload twice
Bug found during testing of #78 

**Steps to Reproduce**
1. Deploy #78 
1. Register LIGO Tale (not needed, but this is a workaround for a separate bug)
1. Login and create a Tale
1. Navigate to `View Tale > Files > Home`
1. Upload a file (in my case `Olaf_invoice_202007.pdf`)
1. Navigate to `View Tale > Files > Workspace`
1. Upload a second file (in my case `bw_polar_2D0000.dat_Slice_z_density.png`)

**Expected Results**
Only the second file is uploaded

**Actual Results**
Two files are created (the one that was uploaded to Home and the second file)

```
$ tree homes
.
├── 5
│   ├── 5f3d4e96d321961376c9190d
│   │   ├── bw_polar_2D0000.dat_Slice_z_density.png  # <- 2nd file
│   │   └── Olaf_invoice_202007.pdf                  # <- 1st file 
│   └── 5f3d4ef7d321961376c9193b
│       ├── environment.yml
│       ├── LOSC_Event_tutorial.ipynb
│       └── readligo.py
├── k
│   └── kowalikk
│       └── Olaf_invoice_202007.pdf
└── temp
```

**NOTES**

### PR DESCRIPTION
**Pull request by bodom0015**
## Problem
Fixes #89 - Files upload twice
Fixes #50 - Adding External Data would return a 400
Fixes #48 - Uploading a file would result in a 400, even though the file appeared to upload successfully
Fixes #41 - Create Folder in Home actually creates the folder in the Tale Workspace
Fixes #37 - Files cannot be removed/downloaded until after refreshing
Fixes #33 - Data registration dropdown should auto-select first result
Fixes #31 - Show previously-added data on select-data-modal
Fixes #30 - Add button should be disabled for read-only Tales
Fixes #28 - External Data shows nothing even after adding a dataset

## Approach
* Remove files from the upload queue after they complete
* Remove extra call to Girder's `/chunk` endpoint - logic for `initUpload` appears to upload correctly (#48)
* Fix `createFolder` to use the same logic for choosing `parentId` as the `uploadFiles` function uses (#41)
* Allow files to be removed and/or downloaded without needing to first refresh the view (#37)
* Auto-select first dropdown result when registering data (#33)
* Rework `select-data-modal` to show Selected Data on the right, as was previously true in the Ember.js dashboard (#31)
* Disable Add button on Run > Files for read-only Tales (#30) 
* Fixed display/fetch of External Data (#28)
* Remove extraneous debug line from Run > Interact view

## How to Test
Prerequisites: at least one Tale created, LIGO Tale

1. Checkout this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Choose a Tale you own and navigate to the Run > Files view
4. Select Home and upload a file
    * You should see that your file is uploaded successfully (#48)
    * You should see that your file has a Last Modified date (this indicates that it is now displaying an `Item`, as expected, and not a `File`)
5. Without refreshing the view, expand the dropdown beside the new file and choose Remove
    * The file should be removed from view (#37)
6. Select Home, create a folder, and refresh the page
    * You should still see the Home nav is still selected
    * Your new folder should still exist (#41)
7. Select External Data, click (+), and choose "Web (DOI or URL)"
    * The data registration modal should open
8. Enter `10.5065/D6862DM8` and search
    * You should see the Humans & Hydrology dataset is found
    * The dataset should be automatically selected in the dropdown (#33)
9. Click Register to close the modal, then click (+) again, and choose "WT Data Catalog"
     * The Select Data modal should open
10. Click the "Humans & Hydrology" dataset to select it, optionally double-click to drill down into it and select individual folders/files, and click "Add Selected. and then 
    * The modal should close
    * You should see the new dataset(s) appear on the right after adding it
11. Click "Save" below to confirm
    * You should see the new dataset(s) appear in the view shortly after the modal closes
    * You should no longer see 400 errors appear in the console when adding the dataset (#50)
12. Click (+) again, and choose "WT Data Catalog" once more
    * You should see the data you've previously added appears on the right by default (#31)
13. Refresh the page
    * You should still see the External Data nav is still selected
    * Your new dataset should still exist on the Tale (#28)
14. Navigate to the Run > Files view for the LIGO Tale
    * You should see the (+) button at the top-right is disabled (#30)

### CODE REVIEW

**Comment by Xarthisius** [comment]
With this pull request #41 is gets reverted. Now when I upload files to Workspace they end up in Home.

**Comment by bodom0015** [comment]
@Xarthisius Good catch! #41 was actually about creating folders, but was likely the same problem as it uses the same logic to determine the `parentId` to use for the operation. There was a typo in the nav it was searching for: `tale_workspaces` should have instead been `tale_workspace`. Thank you for pointing this out :+1:

**Comment by craig-willis** [comment]
Looking good. Testing no. 10 from your list, I got into a very strange state. 

I selected the checkbox for the top level "Humans and Hydrology" dataset and that worked fine.  I went to deselect it and instead select a single item from within the folder and things got weird.  I can uncheck the checkbox, but clicking "Select" results in a new entry for the top level "Humans and Hydrology" folder.

image

image

I can select a single file from within the dataset, but the top level folder remains:

image

Basic test case:
* Select External Data > WT Catalog, check "Humans and Hydrology" folder, click "Select". You should see a single folder in external data.
* Select External Data > WT Catalog, uncheck "Humans and Hydrology" folder, click "Select". You should see two folders in external data.
* Select External Data > WT Catalog, select "Humans and Hydrology" folder, check "usco2005.xls", click "Select". You should see two folders and the file.

**Comment by bodom0015** [comment]
@craig-willis good catch, I've noticed this before but ... I think that modal needs to be reworked quite a bit for #31 as well. The current implementation seems to get confused between what was previously added and what you're trying to add now. The navigation bits can still be preserved, but the additional panel showing a list of what has already been selected may help keep things straight.

**Comment by bodom0015** [comment]
So I have a fix for the `select-data-modal`, but am not sure whether to amend this PR or to make a new one.

@Xarthisius any preference here? Should I make a new PR after this one gets merged, since there's already a lot going on in this one? What do you think?

**Comment by Xarthisius** [comment]
> So I have a fix for the `select-data-modal`, but am not sure whether to amend this PR or to make a new one.
> 
> @Xarthisius any preference here? Should I make a new PR after this one gets merged, since there's already a lot going on in this one? What do you think?

Please add it here and let's review it in a one go.

**Comment by craig-willis** [comment]
Ran into an unexpected behavior while testing the fix. Should "Remove" remove the file/dataset from My External Data (i.e., deselect) or remove from the system catalog? Currently, the "Remove" menu calls DELETE https://girder.local.wholetale.org/api/v1/folder/<id>, effectively removing the dataset/file from Girder. Selected dataset or file still remain selected in "Selected Data"

To repeat:
* Register Humans and Hydrology (10.5065/D6862DM8)
* Select + > My External Data
* Browse to Humans and Hydrology > Add Selected, Save
* H&H appears in external data
* Select "Remove"
* Select "My External Data" -- H&H appears in Selected list
* Go to https://girder.local.wholetale.org/ > Whole Tale Catalog > Whole Tale Catalog -- dataset has been removed from catalog

Something I can't readily repeat -- when I select an individual file in the folder, sometimes the folder is automatically selected for me.  I'll try to get a repeatable test case.

**Comment by bodom0015** [comment]
@craig-willis ahhh, I think this was an oversight on my part. While fixing #30, I think I accidentally made this view writeable.

Will make a new commit to hide the "Remove" option there, is it is in fact deleting the dataset from Girder.

**Comment by Xarthisius** [comment]
@bodom0015 I went through the test plan (and some more). Very nice! I filed issues for some minor things I've found. In particular #89 looks like something you might want to address here, but I'm open to merging this PR as-is and dealing with it in a separate branch. Please let me know what would you prefer.

**Comment by bodom0015** [comment]
Good catch on these! I've pushed a fix for #89 as well as another minor bug I noticed.

Apparently we could not upload the same file twice in a row due to using the onChange handler.. since the selected file technically wasn't changing, the upload would never trigger :D

File: src/app/+run-tale/run-tale/tale-files/tale-files.component.ts [hunk 1]
```diff
@@ -209,6 +209,8 @@ export class TaleFilesComponent implements OnInit, OnChanges {
           });*/
         });
       });
+
+      this.uploadQueue.delete(upload);
```
Comment (bodom0015) [inline src/app/+run-tale/run-tale/tale-files/tale-files.component.ts:213]: I am amazed that this did not trigger a concurrent modification exception.. in other languages, deleting an item from an array that you're looping over is not allowed and likely not advisable here

**Pull request merged**